analyzers/tests/SonarAnalyzer.UnitTest/ResourceTests/GeneratedResourcesTest.cs

```diff
@@ -41,80 +38,60 @@ public class GeneratedResourcesTest
+            var rulesFromResources = SortedRulesFromResources(RspecRelativeFolderPath + "vbnet");           // Unique list
+            var rulesFromTypes = SortedRulesFromTypes(AnalyzerLanguage.VisualBasic).Distinct().ToArray();   // Same ruleId can be in multiple classes (see InvalidCastToInterface)
+            rulesFromResources.Should().Equal(rulesFromTypes);
         }
 
         [TestMethod]
         public void ThereShouldBeRuleDetailsForAllCSharpRuleClasses()
         {
-            var ruleDetailsKeys = RuleDetailBuilder.GetAllRuleDetails(AnalyzerLanguage.CSharp)
-                                                   .Select(rd => rd.Key)
-                                                   .OrderBy(key => key);
-
-            var rulesFromClasses = GetRulesFromClasses(typeof(CSharpSyntaxHelper).Assembly).OrderBy(key => key);
-
-            ruleDetailsKeys.Should().Equal(rulesFromClasses);
+            var ruleDetailsKeys = SortedRulesFromDetailbuilder(AnalyzerLanguage.CSharp);
+            var rulesFromTypes = SortedRulesFromTypes(AnalyzerLanguage.CSharp);
+            ruleDetailsKeys.Should().Equal(rulesFromTypes);
         }
 
         [TestMethod]
         public void ThereShouldBeRuleDetailsForAllVbNetRuleClasses()
         {
-            var ruleDetailsKeys = RuleDetailBuilder.GetAllRuleDetails(AnalyzerLanguage.VisualBasic)
-                                                   .Select(rd => rd.Key)
-                                                   .OrderBy(key => key);
-
-            var rulesFromClasses = GetRulesFromClasses(typeof(VisualBasicSyntaxHelper).Assembly).OrderBy(key => key);
-
-            ruleDetailsKeys.Should().Equal(rulesFromClasses);
+            var ruleDetailsKeys = SortedRulesFromDetailbuilder(AnalyzerLanguage.VisualBasic);
+            var rulesFromTypes = SortedRulesFromTypes(AnalyzerLanguage.VisualBasic);
+            ruleDetailsKeys.Should().Equal(rulesFromTypes);
         }
 
-        private static string[] GetRulesFromClasses(Assembly assembly) =>
-            assembly.GetTypes()
-                    .Where(x => !x.IsAbstract && typeof(SonarDiagnosticAnalyzer).IsAssignableFrom(x) && IsNotUtilityAnalyzer(x))
-                    .SelectMany(GetRuleNamesFromAttributes)
-                    .OrderBy(x => x)
-                    .ToArray();
-
-        private static bool IsNotUtilityAnalyzer(Type arg) =>
-            !typeof(UtilityAnalyzerBase).IsAssignableFrom(arg);
+        private static string[] SortedRulesFromTypes(AnalyzerLanguage language) =>
+            RuleFinder.GetAnalyzers(language)
+                .Where(x => x is not UtilityAnalyzerBase)
+                .SelectMany(x => x.SupportedDiagnostics.Select(desciptor => desciptor.Id))
+                .Distinct() // One class can have the same ruleId multiple times, see S3240, same ruleId can be in multiple classes (see InvalidCastToInterface)
+                .OrderBy(x => x)
+                .ToArray();
```

Comment:
pedantic: the alignment is off

Reply:
This coding style was not documented yet (it's my action item from the last How we work)